### PR TITLE
separate trie from util and make its dependencies into libs

### PR DESCRIPTION
**Pull request by Hawstein**
Part of #6418

Separate trie from util and make its dependencies into libs:
* bytes
* hashdb
* memorydb
* nibbleslice
* nibblevec

### CODE REVIEW

**Comment by Hawstein** [comment]
Finally, all the errors are resolved! And it is ready for review.

**Comment by rphmeier** [comment]
could you rename the trie package to `patricia_trie` (open to suggestions). We can publish it on crates.io once that's done and it's merged.

File: util/trie/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+log = "0.3"
+rand = "0.3"
+ethcore-bytes = { path = "../bytes" }
+ethcore-bigint = { path = "../bigint" }
+hash = { path = "../hash" }
+hashdb = { path = "../hashdb" }
+rlp = { path = "../rlp" }
+nibbleslice = { path = "../nibbleslice" }
+nibblevec = { path = "../nibblevec" }
+triehash = { path = "../triehash" }
```
Comment (rphmeier) [inline util/trie/Cargo.toml]: maybe `triehash` should go in the same crate, within an "ephemeral" module?

**Comment by Hawstein** [comment]
> could you rename the trie package to patricia_trie (open to suggestions). We can publish it on crates.io once that's done and it's merged.

@rphmeier Sure, `patricia-trie ` seems like a reasonable name.

File: util/trie/Cargo.toml (hunk 1 above)
Comment (Hawstein) [reply util/trie/Cargo.toml]: Hmm, I guess we can make it a separated PR if all of we agree to move it to this crate.

**Comment by rphmeier** [comment]
@Hawstein This could use a merge

**Comment by Hawstein** [comment]
@rphmeier Rebase master and rename trie.

Could you review it if you have time? Since it touches lots of files, it will easily has conflicts with other new PRs.

**Comment by Hawstein** [comment]
@gavofyork @debris 

Rebase master. This can be merged without conflicts now.

**Pull request merged**